lib/matplotlib/axes/_axes.py

```diff
@@ -5182,7 +5182,21 @@ def _pcolorargs(funcname, *args, **kw):
             return X, Y, C
 
         if len(args) == 3:
-            X, Y, C = [np.asanyarray(a) for a in args]
+            # Check x and y for bad data...
+            C = np.asanyarray(args[2])
+            X, Y = [cbook.safe_masked_invalid(a) for a in args[:2]]
+            if funcname == 'pcolormesh':
+                if np.ma.is_masked(X) or np.ma.is_masked(Y):
+                    raise TypeError(
+                        'x and y arguments to pcolormesh cannot have '
+                        'non-finite values or be of type '
+                        'numpy.ma.core.MaskedArray with masked values')
+                # if it didn't fail the above it is masked but has no
+                # masked values, so strip the mask for pcolormesh
+                if type(X) is np.ma.core.MaskedArray:
+                    X = X.data
+                if type(Y) is np.ma.core.MaskedArray:
+                    Y = Y.data
```

Comment:
The 4 lines above can be one line:
```python
X, Y = X.data, Y.data
```

Reply:
Not if they are `ndarrays`?

Reply:
I was incorrectly assuming that `safe_masked_invalid` always returned a masked array--and possibly it should do so for consistency, but that is another matter.  Given that it doesn't, you are stuck with the checks, but please use `if isinstance(X, np.ma.core.MaskedArray):` which is simpler and more idiomatic.
It might be worth replacing your comment above these blocks with:
```
# save_masked_invalid() returns an ndarray for dtypes other than floating point
```